### ISSUE TITLE
Aligning canvas control directly over image control

### ISSUE THREAD

**Issue by dumbledad** [Microsoft/Win2D#109]
I am having trouble aligning a canvas control containing a fragment of an image exactly over an image control containing the same image (though re-sized down from a larger copy). I've distilled my problems down into a very simple example.

    <Grid x:Name="MainGrid">
        <Image x:Name="InputItemImage" 
            Source="Assets/Flower720x1280.jpg" 
            Opacity="0.5" 
            Stretch="Uniform"
            HorizontalAlignment="Left" 
            VerticalAlignment="Top"/>
        <canvas:CanvasControl 
            x:Name="MainCanvas"
            CreateResources="MainCanvas_CreateResources"/>
    </Grid>

    private async void MainCanvas_CreateResources(Microsoft.Graphics.Canvas.UI.Xaml.CanvasControl sender, Microsoft.Graphics.Canvas.UI.CanvasCreateResourcesEventArgs args)
    {
        _InputImageBitmap = await CanvasBitmap.LoadAsync(MainCanvas, new Uri("ms-appx:///Assets/Flower360x640.jpg"));
        _InputImageBrush = new CanvasImageBrush(MainCanvas, _InputImageBitmap);

        _DrawingRect = CanvasGeometry.CreateRectangle(MainCanvas, 0, 0, (int)_InputImageBitmap.Size.Width, (int)_InputImageBitmap.Size.Height);

        var opaqueWhite = Color.FromArgb(0xFF, 0xFF, 0xFF, 0xFF);
        var transparentBlack = Color.FromArgb(0x00, 0x00, 0x00, 0x00);
        var canvasRenderTarget = new CanvasRenderTarget(MainCanvas, (float)_InputImageBitmap.Size.Width, (float)_InputImageBitmap.Size.Height);
        using (var ds = canvasRenderTarget.CreateDrawingSession())
        {
            ds.Clear(transparentBlack);
            ds.FillRectangle(0, 0, 200, 400, opaqueWhite);
        }
        _MaskBitmap = (CanvasBitmap)canvasRenderTarget;
        _MaskBrush = new CanvasImageBrush(MainCanvas, _MaskBitmap);

        MainCanvas.Draw += MainCanvas_Draw;
        MainCanvas.Invalidate();
    }

    private void MainCanvas_Draw(Microsoft.Graphics.Canvas.UI.Xaml.CanvasControl sender, Microsoft.Graphics.Canvas.UI.Xaml.CanvasDrawEventArgs args)
    {
        args.DrawingSession.FillGeometry(
            _DrawingRect,
            _InputImageBrush,
            _MaskBrush);
    }

As you can see from this screen shot of the emulator the above code works:

![screenshot1](https://cloud.githubusercontent.com/assets/110564/8270956/3cacba80-17f6-11e5-96ff-c04096a9203a.png)

But I need the underlying image to be centered, and of course just centering the image control leaves the canvas control misaligned:

    <Grid x:Name="MainGrid">
        <Image x:Name="InputItemImage" 
            Source="Assets/Flower720x1280.jpg" 
            Opacity="0.5" 
            Stretch="Uniform"
            HorizontalAlignment="Center" 
            VerticalAlignment="Top"/>
        <canvas:CanvasControl 
            x:Name="MainCanvas"
            CreateResources="MainCanvas_CreateResources"/>
    </Grid>

![screenshot2](https://cloud.githubusercontent.com/assets/110564/8270958/94f63a04-17f6-11e5-83c5-73efbea94d3b.png)

If I position the canvas control it (silently) stops working:

    <Grid x:Name="MainGrid">
        <Image x:Name="InputItemImage" 
            Source="Assets/Flower720x1280.jpg" 
            Opacity="0.5" 
            Stretch="Uniform"
            HorizontalAlignment="Center" 
            VerticalAlignment="Top"/>
        <canvas:CanvasControl 
            x:Name="MainCanvas"
            CreateResources="MainCanvas_CreateResources"
            HorizontalAlignment="Center" 
            VerticalAlignment="Top"/>
    </Grid>

![screenshot3](https://cloud.githubusercontent.com/assets/110564/8270975/d1ceeaa2-17f6-11e5-84ad-1c87a7fcaf78.png)

And I get the same result if I try to align a container for the canvas control:

    <Grid x:Name="MainGrid">
        <Image x:Name="InputItemImage" 
            Source="Assets/Flower720x1280.jpg" 
            Opacity="0.5" 
            Stretch="Uniform"
            HorizontalAlignment="Center" 
            VerticalAlignment="Top"/>
        <Grid x:Name="CanvasGrid"
            HorizontalAlignment="Center" 
            VerticalAlignment="Top">
            <canvas:CanvasControl 
                x:Name="MainCanvas"
                CreateResources="MainCanvas_CreateResources"/>
        </Grid>
    </Grid>

![screenshot4](https://cloud.githubusercontent.com/assets/110564/8270982/24fe820a-17f7-11e5-899d-c17b25ff5c3e.png)

What is the best way to align a fragment in Win2D accurately above another control, especially an image control?

**Comment by shawnhar** [comment]
The problem is that you aren't telling XAML anything about how big you want the CanvasControl to be.  Unlike an element such as as Image, which can determine a sensible size from its Source property, CanvasControl has no way to automatically work out how big you want it to be, so if you don't specify a size it will just report "sure, whatever, I don't care" when XAML is choosing sizes during layout.

In your first version, the CanvasControl is laid out to cover the entire grid, and the grid defaults to covering the entire page, so "I don't care" results in the CanvasControl inheriting the size of the entire display.

But when you tell XAML to center the CanvasControl, to do that it first needs to know the size of the control.  So it asks, and the control says "I don't care", so XAML says "ok, how about zero?" and the control says "sure..." so you end up with a centered but zero sized control.

In short: to do layout on a control, you need to specify how big you want it to be, eg. by setting MinWidth and MinHeight properties.

**Comment by dumbledad** [comment]
Thanks Sawn - that's useful.

I ended up doing it in code not in XAML:

    private void MainCanvas_Draw(Microsoft.Graphics.Canvas.UI.Xaml.CanvasControl sender, Microsoft.Graphics.Canvas.UI.Xaml.CanvasDrawEventArgs args)
    {
        DrawToBitmap();
        var width = (int)InputItemImage.ActualWidth;
        var height = (int)InputItemImage.ActualHeight;
        var gridWidth = MainGrid.ActualWidth;
        var xOffset = (int)((gridWidth - width) / 2);
        var yOffset = 0;
        args.DrawingSession.DrawImage(
            (CanvasBitmap)_DrawingRenderTarget,
            new Rect(xOffset, yOffset, width, height),
            new Rect(0, 0, _DrawingRenderTarget.Size.Width, _DrawingRenderTarget.Size.Height),
            1f,
            CanvasImageInterpolation.NearestNeighbor,
            CanvasComposite.Copy);
    }

    private void DrawToBitmap()
    {
        using (var ds = _DrawingRenderTarget.CreateDrawingSession())
        {
            ds.Clear(Color.FromArgb(0x00, 0x00, 0x00, 0x00));
            ds.FillGeometry(
                _DrawingRect,
                _InputImageBrush,
                _MaskBrush);
        }
    }
